### PR TITLE
Add multilevel crop and resize functions

### PR DESCRIPTION
**Pull request by syiming**
# Description

> :memo: Add funtions to perform multilevel crop and resize. 
>  
> * Add multilevel_native_crop_and_resize.
> * Add multilevel_matmul_crop_and_resize
> * Change faster rcnn meta arch function that use _crop_and_resize_fn

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [ ] TensorFlow 2 migration
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] A new research paper code implementation
- [x] Other (Add functions support faster rcnn fpn feature extractor)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Simple unittest for both functions.

**Test Configuration**:

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: research/object_detection/meta_architectures/faster_rcnn_meta_arch.py [hunk 1]
```diff
@@ -1963,7 +1963,7 @@ def _compute_second_stage_input_feature_maps(self, features_to_crop,
     second stage box classifier for each proposal.
 
     Args:
-      features_to_crop: A float32 tensor with shape
+      features_to_crop: A list of float32 tensor with shape
```
Comment (tombstone) [inline research/object_detection/meta_architectures/faster_rcnn_meta_arch.py:1966]: Leave this as a single tensor in this PR to not break the calling code.

File: research/object_detection/meta_architectures/faster_rcnn_meta_arch.py [hunk 2]
```diff
@@ -1973,9 +1973,15 @@ def _compute_second_stage_input_feature_maps(self, features_to_crop,
     Returns:
       A float32 tensor with shape [K, new_height, new_width, depth].
     """
+    num_levels = len(features_to_crop)
```
Comment (tombstone) [inline research/object_detection/meta_architectures/faster_rcnn_meta_arch.py:1976]: Just before this line set 
features_to_crop = [features_to_crop]

File: research/object_detection/utils/ops.py [hunk 3]
```diff
@@ -32,8 +32,12 @@
 from object_detection.utils import static_shape
 
 
+multilevel_matmul_crop_and_resize = \
```
Comment (tombstone) [inline research/object_detection/utils/ops.py:35]: No need to add these symbols here. Directly import spatial_ops in model_builder.py and use these functions.

File: research/object_detection/utils/spatial_transform_ops.py [hunk 4]
```diff
@@ -411,6 +411,28 @@ def multilevel_roi_align(features, boxes, box_levels, output_size,
     return features_per_box
 
 
+def multilevel_native_crop_and_resize(images, boxes, box_levels,
+                                      crop_size, scope=None):
+  """Multilevel native crop and resize.
+
+  Same as `multilevel_matmul_crop_and_resize` but uses tf.image.crop_and_resize.
```
Comment (tombstone) [inline research/object_detection/utils/spatial_transform_ops.py:418]: describe input arguments and return values in docstring.

File: research/object_detection/utils/spatial_transform_ops.py [hunk 5]
```diff
@@ -431,6 +453,49 @@ def get_box_inds(proposals):
...
+      cropped image patches are resized to this size. The aspect ratio of the
+      image content is not preserved. Both `crop_height` and `crop_width` need
+      to be positive.
+    extrapolation_value: a float value to use for extrapolation.
+    scope: A name for the operation (optional).
+
+  Returns:
+    A 5-D tensor of shape `[batch, num_boxes, crop_height, crop_width, depth]`
+  """
+  with tf.name_scope(scope, 'MatMulCropAndResize'):
```
Comment (tombstone) [inline research/object_detection/utils/spatial_transform_ops.py:488]: change to MultiLevelMatMulCropAndResize

File: research/object_detection/utils/spatial_transform_ops.py [hunk 6]
```diff
@@ -411,6 +411,28 @@ def multilevel_roi_align(features, boxes, box_levels, output_size,
     return features_per_box
 
 
+def multilevel_native_crop_and_resize(images, boxes, box_levels,
+                                      crop_size, scope=None):
+  """Multilevel native crop and resize.
+
+  Same as `multilevel_matmul_crop_and_resize` but uses tf.image.crop_and_resize.
+  """
```
Comment (tombstone) [inline research/object_detection/utils/spatial_transform_ops.py:419]: add a name scope here like in the other function.

File: research/object_detection/utils/spatial_transform_ops.py [hunk 7]
```diff
@@ -411,6 +411,54 @@ def multilevel_roi_align(features, boxes, box_levels, output_size,
...
+                                      crop_size, scope=None):
+  """Multilevel native crop and resize.
+
+  Same as `multilevel_matmul_crop_and_resize` but uses tf.image.crop_and_resize.
+
+  Args:
+    images: A list of 4-D tensor of shape
+      [batch, image_height, image_width, depth] representing features of
+      different size.
+    boxes: A `Tensor` of type `float32` or 'bfloat16'.
```
Comment (tombstone) [inline research/object_detection/utils/spatial_transform_ops.py:424]: remove bfloat16 as this is a cpu only op.

File: research/object_detection/meta_architectures/faster_rcnn_meta_arch.py [hunk 8]
```diff
@@ -1973,9 +1973,16 @@ def _compute_second_stage_input_feature_maps(self, features_to_crop,
...
     """
+    features_to_crop = [features_to_crop]
+    num_levels = len(features_to_crop)
+    box_levels = None
+    if num_levels != 1:
+      # If there are mutiple levels to select, get the box levels
+      box_levels = ops.fpn_feature_levels(num_levels, num_levels - 1,
+                                          1.0/224, proposal_boxes_normalized)
     cropped_regions = self._flatten_first_two_dimensions(
         self._crop_and_resize_fn(
```
Comment (tombstone) [inline research/object_detection/meta_architectures/faster_rcnn_meta_arch.py:1984]: Please update other references of _crop_and_resize_fn in faster_rcnn_meta_arch.py

Comment (syiming) [reply research/object_detection/meta_architectures/faster_rcnn_meta_arch.py:1959]: Hi, I updated _crop_and_resize_fn in both faster_rcnn_meta_arch.py and context_rcnn_meta_arch.py. Currently, all the tests except the ones related to pycocotools/cocoeval.py and used TPU are passed.

**Comment by syiming** [comment]
Hi @alexyangmie, this is the pr for multilevel crop and resize functions. Thank you!

**Pull request merged**